File: scala/P09.scala
// P09 Pack consecutive duplicates of list elements into sublists.

// Recursion with "span"
def f1[T]: List[T] => List[List[T]] = {
  case Nil => Nil
  case h :: t =>
    val (a, b) = t.span(h ==)
    (h :: a) :: f1(b)
}

File: haskell/p09.hs
-- P09 Pack consecutive duplicates of list elements into sublists.

module P09 where

import Data.List

-- Built-in function
f0 :: Eq a => [a] -> [[a]]
f0 = group

-- Recursion with "span"
f1 :: Eq a => [a] -> [[a]]
f1 [] = []
f1 (x:xs) = let (a, b) = span (== x) xs in (x : a) : f1 b